File: Java Barcode Scanner ME/app/src/main/java/de/dimedis/mobileentry/db/model/TicketHistoryItem.java
package de.dimedis.mobileentry.db.model;

import android.content.ContentValues;
import android.database.Cursor;

import androidx.annotation.NonNull;

import com.google.gson.Gson;

import de.dimedis.lmlib.AccessImpl;
import de.dimedis.lmlib.ActionJson;
import de.dimedis.mobileentry.db.tables.TicketsHistoryTable;
import de.dimedis.mobileentry.model.Action;
import de.dimedis.mobileentry.model.LocalModeHelper;
import de.dimedis.mobileentry.model.PerfData;

public class TicketHistoryItem {

    private Integer columnId;
    private PerfData mPerfData;
    private ActionJson mActionJson;
    private AccessImpl mAccessImpl;

    public PerfData getmPerfData() {
        return mPerfData;
    }

    public void setmPerfData(PerfData mPerfData) {
        this.mPerfData = mPerfData;
    }

    public AccessImpl getAccessImpl() {
        return mAccessImpl;
    }

    public void setAccessImpl(AccessImpl mAccessImpl) {
        this.mAccessImpl = mAccessImpl;
        setActionJson(mAccessImpl);
    }

    public TicketHistoryItem() {
    }

    public String denyReason;

    public String getDenyReason() {
        return denyReason;
    }

    public void setDenyReason(String denyReason) {
        this.denyReason = denyReason;
    }

    void initFieldFromPref() {
        mPerfData = new PerfData().initFromPref();
    }

    public static TicketHistoryItem performEntry(@NonNull String ticketCode) {
        TicketHistoryItem item = new TicketHistoryItem();
        item.initFieldFromPref();
        item.setAccessImpl((AccessImpl) LocalModeHelper.performEntry(item.getmPerfData(), ticketCode));
        return item;
    }

    public static TicketHistoryItem performCheckout(@NonNull String ticketCode) {
        TicketHistoryItem item = new TicketHistoryItem();
        item.initFieldFromPref();
        item.setAccessImpl((AccessImpl) LocalModeHelper.performCheckout(item.getmPerfData(), ticketCode));
        return item;
    }

    public ContentValues toContentValues() {
        ContentValues contentValues = new ContentValues();
        contentValues.put(TicketsHistoryTable.COLUMN_LANG, mPerfData.getLang());
        contentValues.put(TicketsHistoryTable.COLUMN_COMM_KEY, mPerfData.getCommKey());
        contentValues.put(TicketsHistoryTable.COLUMN_DEVICESUID, mPerfData.getDeviceSuid());
        contentValues.put(TicketsHistoryTable.COLUMN_USERSESSION, mPerfData.getUserSession());
        contentValues.put(TicketsHistoryTable.COLUMN_USERSUID, mPerfData.getUserSuid());
        contentValues.put(TicketsHistoryTable.COLUMN_USERNAME, mPerfData.getUserName());
        contentValues.put(TicketsHistoryTable.COLUMN_FAIR, mPerfData.getFair());
        contentValues.put(TicketsHistoryTable.COLUMN_BORDER, mPerfData.getBorder());
        contentValues.put(TicketsHistoryTable.COLUMN_BODY, getActionJson());
        return contentValues;
    }

    public static TicketHistoryItem fromCursor(Cursor cursor) {
        return new TicketHistoryItem(cursor);
    }

    public TicketHistoryItem(Cursor cursor) {
        setColumnId(cursor.getInt(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_ID)));
        mPerfData = new PerfData();
        mPerfData.setLang(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_LANG)));
        mPerfData.setCommKey(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_COMM_KEY)));
        mPerfData.setDeviceSuid(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_DEVICESUID)));
        mPerfData.setUserSession(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERSESSION)));
        mPerfData.setUserSuid(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERSUID)));
        mPerfData.setUserName(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERNAME)));
        mPerfData.setFair(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_FAIR)));
        mPerfData.setBorder(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_BORDER)));
        setActionJson(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_BODY)));
    }

    public static TicketHistoryItem fromAction(String userSUID, Action action) {
        TicketHistoryItem item = new TicketHistoryItem();
        PerfData pf = new PerfData();
        item.setmPerfData(pf);
        ActionJson actionJson = new ActionJson();
        item.setAction(actionJson);
        actionJson.setTicketCode(action.ticketCode);
        actionJson.setUserSuid(userSUID);
        actionJson.setDeviceSuid(action.deviceSuid);
        actionJson.setDeviceName(action.deviceName);
        actionJson.setType(action.type);
        actionJson.setMessage(action.message);
        actionJson.setTimeStamp(action.timestamp);

        actionJson.setDateText(action.dateText);
        actionJson.setTimeText(action.timeText);
        actionJson.setFair(action.location);
        //--\
        item.setDenyReason("");
        return item;
    }

    public Action toAction() {
        return new Action(mActionJson.getTimestamp(), getDateText(), mActionJson.getTimeText(), mActionJson.getDeviceSuid(),
                mActionJson.getDeviceName(), mActionJson.getType(), mActionJson.getMessage(), mActionJson.getTicketCode(), null);
    }

    public String getActionJson() {
        return new Gson().toJson(mActionJson);
    }

    public ActionJson getAction() {
        return mActionJson;
    }

    public void setAction(ActionJson actionJson) {
        mActionJson = actionJson;
    }

    public void setActionJson(AccessImpl access) {
        mActionJson = new ActionJson(mPerfData.getUserSuid(), access);
    }

    public void setActionJson(String json) {
        mActionJson = new Gson().fromJson(json, ActionJson.class);
    }

    public static TicketHistoryItem createDummyItem() {
        TicketHistoryItem item = new TicketHistoryItem();
        item.setAction(new ActionJson());
        return item;
    }


    public Integer getColumnId() {
        return columnId;
    }

    public void setColumnId(Integer columnId) {
        this.columnId = columnId;
    }


    public String getDateText() {
        return mActionJson.getDateText();
    }

}

File: Kotlin Migration/NTMobileEntry/app/src/main/java/de/dimedis/mobileentry/db/model/TicketHistoryItem.kt
package de.dimedis.mobileentry.db.model

import android.content.ContentValues
import android.database.Cursor
import com.google.gson.Gson
import de.dimedis.lmlib.AccessImpl
import de.dimedis.lmlib.ActionJson
import de.dimedis.mobileentry.db.tables.TicketsHistoryTable
import de.dimedis.mobileentry.model.Action
import de.dimedis.mobileentry.model.LocalModeHelper
import de.dimedis.mobileentry.model.PerfData

class TicketHistoryItem {
    var columnId: Int? = null
    private var mPerfData: PerfData? = null
    private var mActionJson: ActionJson? = null
    private var mAccessImpl: AccessImpl? = null
    fun getmPerfData(): PerfData? {
        return mPerfData
    }

    fun setmPerfData(mPerfData: PerfData?) {
        this.mPerfData = mPerfData
    }

    var accessImpl: AccessImpl?
        get() = mAccessImpl
        set(mAccessImpl) {
            this.mAccessImpl = mAccessImpl
//            actionJson = mAccessImpl.toString()
        }

    constructor()

    var denyReason: String? = null
    fun initFieldFromPref() {
        mPerfData = PerfData().initFromPref()
    }

    fun toContentValues(): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(TicketsHistoryTable.COLUMN_LANG, mPerfData?.lang)
        contentValues.put(TicketsHistoryTable.COLUMN_COMM_KEY, mPerfData?.commKey)
        contentValues.put(TicketsHistoryTable.COLUMN_DEVICESUID, mPerfData?.deviceSuid)
        contentValues.put(TicketsHistoryTable.COLUMN_USERSESSION, mPerfData?.userSession)
        contentValues.put(TicketsHistoryTable.COLUMN_USERSUID, mPerfData?.userSuid)
        contentValues.put(TicketsHistoryTable.COLUMN_USERNAME, mPerfData?.userName)
        contentValues.put(TicketsHistoryTable.COLUMN_FAIR, mPerfData?.fair)
        contentValues.put(TicketsHistoryTable.COLUMN_BORDER, mPerfData?.border)
        return contentValues
    }

    constructor(cursor: Cursor) {
        columnId = cursor.getInt(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_ID))
        mPerfData = PerfData()
        mPerfData?.lang = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_LANG))
        mPerfData?.commKey = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_COMM_KEY))
        mPerfData?.deviceSuid = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_DEVICESUID))
        mPerfData?.userSession = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERSESSION))
        mPerfData?.userSuid = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERSUID))
        mPerfData?.userName = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_USERNAME))
        mPerfData?.fair = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_FAIR))
        mPerfData?.border = cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_BORDER))
        setActionJson(cursor.getString(cursor.getColumnIndexOrThrow(TicketsHistoryTable.COLUMN_BODY)));
    }

    fun toAction(): Action {
        return Action(
            mActionJson!!.getTimestamp(), getDateText(),
            mActionJson!!.getTimeText(),
            mActionJson!!.getDeviceSuid(),
            mActionJson!!.getDeviceName(),
            mActionJson!!.getType(),
            mActionJson!!.getMessage(),
            mActionJson!!.getTicketCode(), null)
    }

    fun getAction(): ActionJson? {
        return mActionJson
    }

    fun setAction(actionJson: ActionJson?) {
        mActionJson = actionJson
    }

    fun setActionJson(access: AccessImpl?) {
        mActionJson = access?.let { ActionJson(mPerfData?.userSuid, it) }
    }

    fun setActionJson(json: String?) {
        mActionJson = Gson().fromJson(json, ActionJson::class.java)
    }

    fun getDateText(): String? {
        return mActionJson?.getDateText()
    }

    companion object {
        fun performEntry(ticketCode: String): TicketHistoryItem {
            val item = TicketHistoryItem()
            item.initFieldFromPref()
            item.accessImpl = item.getmPerfData()?.let { LocalModeHelper.performEntry(it, ticketCode) } as AccessImpl
            return item
        }

        fun performCheckout(ticketCode: String): TicketHistoryItem {
            val item = TicketHistoryItem()
            item.initFieldFromPref()
            item.accessImpl = item.getmPerfData()?.let { LocalModeHelper.performCheckout(it, ticketCode) } as AccessImpl
            return item
        }

        fun fromCursor(cursor: Cursor): TicketHistoryItem {
            return TicketHistoryItem(cursor)
        }

        fun fromAction(userSUID: String?, action: Action): TicketHistoryItem {
            val item = TicketHistoryItem()
            val pf = PerfData()
            item.setmPerfData(pf)
            val actionJson = ActionJson()
            item.setAction(actionJson)
            actionJson.setTicketCode(action.ticketCode)
            actionJson.userSuid = (userSUID)
            actionJson.setDeviceSuid(action.deviceSuid)
            actionJson.setDeviceName(action.deviceName)
            actionJson.setType(action.type)
            actionJson.setMessage(action.message)
            actionJson.setTimeStamp(action.timestamp)
            actionJson.setDateText(action.dateText)
            actionJson.setTimeText(action.timeText)
            actionJson.setFair(action.location)
            //--\
            item.denyReason = ""
            return item
        }

        fun createDummyItem(): TicketHistoryItem {
            val item = TicketHistoryItem()
            item.setAction(ActionJson())
            return item
        }
    }
}